src/api/server.rs

```diff
@@ -0,0 +1,134 @@
+use super::{handler, schema};
+use async_graphql::{
+    http::{playground_source, GraphQLPlaygroundConfig},
+    QueryBuilder,
+};
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::{IpAddr, SocketAddr};
+use tokio::sync::oneshot;
+use warp::{http::Response, Filter};
+
+pub struct Server {
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: oneshot::Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<oneshot::Receiver<()>>,
+
+    /// Enables the playground
+    playground: bool,
+}
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Self {
+        let (trigger_cancel, cancel_signal) = oneshot::channel::<()>();
+
+        Server {
+            address,
+            trigger_cancel,
+            cancel_signal: Some(cancel_signal),
+            playground: true,
+        }
+    }
+
+    pub fn set_playground(mut self, enable: bool) -> Self {
+        self.playground = enable;
+        self
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> IpAddr {
+        self.address.ip()
+    }
+
+    /// String representation of the bound port
+    pub fn port(&self) -> u16 {
+        self.address.port()
+    }
+
+    /// Stops the running API server
+    pub fn stop(self) {
+        let _ = self.trigger_cancel.send(());
+    }
+
+    /// Run the API server
+    pub async fn run(mut self) -> Self {
+        let rx = self
+            .cancel_signal
+            .take()
+            .expect("Run can only be called once");
+
+        // Build the GraphQL schema
+        let schema = schema::build_schema().finish();
+
+        // Routes...
+
+        // Health route
+        let health_route = warp::path("health").and_then(handler::health);
+
+        // GraphQL POST handler
+        let graphql_post = async_graphql_warp::graphql(schema.clone()).and_then(
+            |(schema, builder): (_, QueryBuilder)| async move {
+                let resp = builder.execute(&schema).await;
+                Ok::<_, Infallible>(GQLResponse::from(resp))
+            },
+        );
+
+        // GraphQL playground
+        let enable_playground = self.playground;
+        let graphql_playground = warp::path("playground").map(move || {
+            if enable_playground {
+                Response::builder()
+                    .header("content-type", "text/html")
+                    .body(playground_source(
+                        GraphQLPlaygroundConfig::new("/graphql").subscription_endpoint("/graphql"),
+                    ))
+            } else {
+                Response::builder()
```

Comment:
Would this work?
```rust
let mut routes = balanced_or_tree!(...); // non-conditional routes
if enable_playground {
  routes = routes.or(graphql_playground);
}
routes.with(...) // the rest 0f the initialization
```

Reply:
This is the error:

```rust
error[E0308]: mismatched types
   --> src/api/server.rs:121:22
    |
106 |               let graphql_playground = warp::path("playground").map(move || {
    |  ___________________________________________________________________-
107 | |                 if enable_playground {
108 | |                     Response::builder()
109 | |                         .header("content-type", "text/html")
...   |
118 | |                 }
119 | |             });
    | |_____________- the found closure
120 | 
121 |               routes = routes.or(graphql_playground);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `warp::filter::boxed::BoxedFilter`, found struct `warp::filter::or::Or`
    | 
   ::: /Users/leebenson/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-warp-1.17.3/src/lib.rs:162:19
    |
162 |   ) -> BoxedFilter<(impl Reply,)>
    |                     ----------
    |                     |
    |                     one of the expected opaque types
    |                     one of the found opaque types
    | 
   ::: src/api/handler.rs:11:33
    |
11  |   pub async fn health() -> Result<impl Reply, Rejection> {
    |                                   ----------
    |                                   |
    |                                   one of the expected opaque types
    |                                   one of the found opaque types
    |
    = note: expected struct `warp::filter::or::Or<warp::filter::or::Or<warp::filter::boxed::BoxedFilter<_>, warp::filter::boxed::BoxedFilter<(impl warp::reply::Reply,)>>, warp::filter::or::Or<warp::filter::boxed::BoxedFilter<(async_graphql_warp::GQLResponse,)>, warp::filter::boxed::BoxedFilter<(std::result::Result<http::response::Response<std::string::String>, http::error::Error>,)>>>`
               found struct `warp::filter::or::Or<warp::filter::or::Or<warp::filter::or::Or<warp::filter::boxed::BoxedFilter<_>, warp::filter::boxed::BoxedFilter<(impl warp::reply::Reply,)>>, warp::filter::or::Or<warp::filter::boxed::BoxedFilter<(async_graphql_warp::GQLResponse,)>, warp::filter::boxed::BoxedFilter<(std::result::Result<http::response::Response<std::string::String>, http::error::Error>,)>>>, warp::filter::map::Map<warp::filters::path::Exact<warp::filters::path::internal::Opaque<&str>>, [closure@src/api/server.rs:106:67: 119:14 enable_playground:_]>>`
```

Reply:
I see. The issue here is that the `.or` combinator returns a new value with a different type, that wraps the original. So, in case `if` executes the type of the `routes.or(graphql_playground)` would be different from what rust had initially eluded at `let route`, and thus not be assignable.

We can work around it by using `Box` with a trait object in it - this way the types will simply be homomorphic `Box<dyn Filter>` in all branches, thus resolving the type conflict.

Warp offers some neat helpers, so what we should be able to do is:

```rust
let mut routes = balanced_or_tree!(...).boxed(); // non-conditional routes
if enable_playground {
  routes = routes.or(graphql_playground).boxed();
}
routes.with(...) // the rest 0f the initialization
```

Reply:
Apparently there's even no need to call `.boxed()` for `balanced_or_tree!(...)` (cause macro already does that internally) - just for `routes.or(graphql_playground).boxed()`.

Reply:
I believe`.boxed()` is only called on debug builds, deferring to dynamic dispatch:

```rust
#[cfg(debug_assertions)]
macro_rules! debug_boxed {
    ($x:expr) => {
        ::warp::Filter::boxed($x)
    };
}

#[cfg(not(debug_assertions))]
macro_rules! debug_boxed {
    ($x:expr) => {
        $x
    };
}
```

Doesn't boxing it make routing slightly slower? I'm not sure if the overhead is enough to make a difference either way. I'd prefer to avoid the route if the performance penalty is negligible.

Reply:
This is a good question, and I really don't know the answer. It seems like in our case the effect of it to performance won't matter, but, as always, this needs micro benches to say for sure.
There seems to be an alternative to dynamic dispatch in wrap though - the `unify` call. https://docs.rs/warp/0.2.4/warp/trait.Filter.html#method.unify

I'm not sure, but it looks like it could be used like this:

```rust
let routes = balanced_or_tree!(...); // non-conditional routes
let routes = if enable_playground {
  routes.or(graphql_playground).unify()
} else {
  routes.unify()
};
```

Reply:
👍 Thanks, I'll take a look at that!

Reply:
I updated the post above - just figured it's probably not for our use case. Also, see the commend re `boxed` :smile:

Reply:
Haven't trouble getting the boxed types to work:

```rust
error[E0308]: mismatched types
   --> src/api/server.rs:63:18
    |
63  |         routes = routes.or(playground).boxed();
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected opaque type, found enum `warp::generic::Either`
```

If you have the free cycles, would you mind updating this branch with a demo please? I'm getting lost in nested types.

Reply:
Sure!

Reply:
Take a look at https://github.com/timberio/vector/pull/3618

Reply:
Thanks @MOZGIII, will take a look now